docs: add structure to 12-factor reference pages (#2345)
canonical/charmcraft: docs/reuse/reference/extensions/integrations.rst

> Relation endpoints
> ------------------

Reviewer: ```suggestion
Relations
---------
```

Relations are endpoints. https://documentation.ubuntu.com/juju/3.6/reference/relation/

        interface: ingress
        limit: 1

In addition to these relations, in each ``provides`` and ``requires``

> block you may specify further relations endpoints, to integrate with

Reviewer: ```suggestion
block you may specify further relations, to integrate with
```

Following the heading.

the following charms:

.. list-table::

>   :widths: 30 30

Reviewer: Minute detail. These numbers form a ratio, so this is equivalent to `1:1`.

  :header-rows: 1

  * - Relation
    - Endpoint definition

>   * - Ingress: `traefik <https://charmhub.io/traefik-k8s>`__ and `nginx

Reviewer: I think we could do without the colons.

      ingress integrator <https://charmhub.io/nginx-ingress-integrator>`__
    - Already available in the charm
  * - MySQL: `machine <https://charmhub.io/mysql>`__ and

>       `k8s <https://charmhub.io/mysql-k8s>`__ charm

Reviewer: For the generic descriptor, not the charm name, should it be "Kubernetes charm"? The reference suggests so. https://documentation.ubuntu.com/juju/3.6/reference/charm/#kubernetes

    - .. code-block:: yaml

          requires:
            mysql:
              interface: mysql_client
              optional: True
              limit: 1

  * - PostgreSQL: `machine <https://charmhub.io/postgresql>`__ and
      `k8s <https://charmhub.io/postgresql-k8s>`__ charm
    - .. code-block:: yaml

          requires:
            postgresql:
              interface: postgresql_client
              optional: True
              limit: 1

>   * - `MongoDB <https://charmhub.io/mongodb>`__

Reviewer: Add "charm" to the end of each of these?

    - .. code-block:: yaml

         requires:
           rabbitmq:
             interface: rabbitmq
             optional: True
             limit: 1

  * - `Tempo <https://charmhub.io/topics/charmed-tempo-ha>`__
    - .. code-block:: yaml

          requires:
            tracing:
              interface: tracing
              optional: True
              limit: 1

  * - `SMTP <https://charmhub.io/smtp-integrator>`__
    - .. code-block:: yaml

          requires:
            smtp:
              interface: smtp
              optional: True
              limit: 1

  * - `OpenFGA <https://charmhub.io/openfga-k8s>`__
    - .. code-block:: yaml

          requires:
            openfga:
              interface: openfga
              optional: True
              limit: 1

.. note::

    The key ``optional`` with value ``False`` means that the charm will
    get blocked and stop the services if the integration is not provided.

To add one of these relations, e.g., PostgreSQL, in the
project file, include the appropriate ``requires`` block and
integrate with |juju_integrate_postgresql| as usual.

Environment variables
~~~~~~~~~~~~~~~~~~~~~

> After the relation has been established, environment variables will be
> provided to your |framework| app:

Reviewer: ```suggestion
Each relation adds its own environment variables to your |framework| app. Some
must be set for the relation to function.
```

We're combining both remote past and future tense in the same sentence, which is a peeve of mine. I also added a hint about what "required" means.

.. list-table::
  :widths: 20 40
  :header-rows: 1

  * - Relation

>     - Available environment variables

Reviewer: ```suggestion
    - Environment variables
```

I'm not sure "Available" is right, given that some are required.

        - ``S3_STORAGE_CLASS``
        - ``S3_ENDPOINT``
        - ``S3_PATH``
        - ``S3_API_VERSION``
        - ``S3_URI_STYLE``
        - ``S3_ADDRESSING_STYLE``
        - ``S3_ATTRIBUTES``
        - ``S3_TLS_CA_CHAIN``
  * - RabbitMQ
    -
        - ``RABBITMQ_CONNECT_STRING``
        - ``RABBITMQ_SCHEME``
        - ``RABBITMQ_NETLOC``
        - ``RABBITMQ_PATH``
        - ``RABBITMQ_PARAMS``
        - ``RABBITMQ_QUERY``
        - ``RABBITMQ_FRAGMENT``
        - ``RABBITMQ_USERNAME``
        - ``RABBITMQ_PASSWORD``
        - ``RABBITMQ_HOSTNAME``
        - ``RABBITMQ_PORT``
        - ``RABBITMQ_VHOST``
  * - Tracing
    -
        - ``OTEL_EXPORTER_OTLP_ENDPOINT``
        - ``OTEL_SERVICE_NAME``
  * - SMTP
    -
        - ``SMTP_HOST``
        - ``SMTP_PORT``
        - ``SMTP_USER``
        - ``SMTP_PASSWORD_ID``
        - ``SMTP_AUTH_TYPE``
        - ``SMTP_TRANSPORT_SECURITY``
        - ``SMTP_DOMAIN``
  * - OpenFGA
    -
        - ``FGA_STORE_ID``
        - ``FGA_TOKEN``
        - ``FGA_GRPC_API_URL``
        - ``FGA_HTTP_API_URL``

The environment variable |base_url| provides the Ingress URL
for an Ingress relation or the Kubernetes service URL if there is no

> Ingress relation.

Reviewer: As this comes after a large table, we might need be concerned about readers missing it. Should we move it to the introduction?

Author: Sure, I'll move it to the introduction
